In [50]:
import pandas as pd
import numpy as np
import seaborn as sns
import scipy.stats as ss
import datetime

# Exam 1 - Coding Portion

In this section, please answer the following questions and show your work in the notebook. For each response, please check carefully what it asks, and give a response matching that. Also, please clean it up a bit so things are readable, I should be able to find what you found and follow. 

Grading breakdown:
<ul>
<li> Can I read and follow this code? 5 points. </li>
<li> Q1 - Generalize to population - 5 points. </li>
<li> Q2 - Correlation - 3 points. </li>
<li> Q3 - Multiple Groups - 5 points. </li>
<li> Q4 - Store Problem - 7 points. </li>
</ul>

Other than what's specifically instructed, you can use/import pretty much anything and take any approach you want. Just be sure to explain what you're doing and why, and make it clear what your final answers are.

## Submission

To submit, just commit and push your changes to your GitHub repository, as normal. There's no auotmated testing or anything, if yours is visible online on the repository page, you're good. 

## Data Preparation - Don't Edit This Section

In [51]:
data_uae = pd.read_csv("data/uae_real_estate_2024.csv")
data_uae = data_uae[data_uae['price'] <= 5000000]
data_uae['verified'] = data_uae['verified'].astype(bool)


data_uae["sq_ft"] = data_uae["sizeMin"].apply(lambda x: int(x.split(" ")[0].replace(",", "")) if pd.notnull(x) else np.nan)
data_uae["description_length"] = data_uae["description"].apply(lambda x: len(x) if pd.notnull(x) else 0)
data_uae["log_price"] = np.log(data_uae["price"])
data_uae.head()


data_uae.head()

title  \
0               Great ROI I High Floor I Creek View   
2          Green Belt | Corner Unit | Spacious Plot   
3          2BR+Study | Near Pool and Park | Private   
4            Vacant | Well Maintained | Area Expert   
6  High Floor | Motivated Seller | Full Marina View   

                                      displayAddress bathrooms bedrooms  \
0               Binghatti Canal, Business Bay, Dubai         3        2   
2            La Rosa 6, Villanova, Dubai Land, Dubai         3        4   
3                     Springs 15, The Springs, Dubai         3        2   
4                Noor Townhouses, Town Square, Dubai         3        3   
6  Vida Residences Dubai Marina, Dubai Marina, Dubai         3        2   

                addedOn                  type    price  verified  \
0  2024-08-14T12:02:53Z  Residential for Sale  2500000      True   
2  2024-08-14T06:24:28Z  Residential for Sale  3600000      True   
3  2024-08-15T06:07:22Z  Residential for Sale  2999999      True   
4  2024-08-09T08:28:59Z  Residential for Sale  2449999      True   
6  2024-08-16T05:03:07Z  Residential for Sale  4980000      True   

  priceDuration    sizeMin furnishing  \
0          sell  1323 sqft         NO   
2          sell  2324 sqft         NO   
3          sell  1647 sqft         NO   
4          sell  2105 sqft         NO   
6          sell  1095 sqft         NO   

                                         description  sq_ft  \
0  MNA Properties is delighted to offer this apar...   1323   
2  Treo Homes is very pleased to be bringing to t...   2324   
3  2 Bedrooms + Study | Near Pool &amp; Park | Pr...   1647   
4  -Type 1\n-3 Bed+Maid\n-Close To Amenities\n-BU...   2105   
6  Property Features:\n* 2 Bedrooms\n* Unfurnishe...   1095   

   description_length  log_price  
0                1272  14.731801  
2                1102  15.096444  
3                1025  14.914123  
4                1026  14.711598  
6                1286  15.420940

In [52]:
data_uae.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3961 entries, 0 to 5057
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   title               3961 non-null   object 
 1   displayAddress      3961 non-null   object 
 2   bathrooms           3882 non-null   object 
 3   bedrooms            3881 non-null   object 
 4   addedOn             3961 non-null   object 
 5   type                3961 non-null   object 
 6   price               3961 non-null   int64  
 7   verified            3961 non-null   bool   
 8   priceDuration       3961 non-null   object 
 9   sizeMin             3961 non-null   object 
 10  furnishing          3959 non-null   object 
 11  description         3961 non-null   object 
 12  sq_ft               3961 non-null   int64  
 13  description_length  3961 non-null   int64  
 14  log_price           3961 non-null   float64
dtypes: bool(1), float64(1), int64(3), object(10)
memory usage: 4

## Price of Population

Our data has a sample of real estate listings in the UAE. Use this sample to estimate the mean price as well as a 95% confidence interval for the mean price of all real estate listings in the UAE.

<b>If needed, use a value of 875000 for the population count of number of homes in the UAE.</b>

In [53]:
# Generalize 

# Calculate the mean and standard deviation
sample_mean = data_uae['price'].mean()
sample_std = data_uae['price'].std()
n = data_uae.shape[0]

# Compute standard error
standard_error = sample_std / np.sqrt(n)

# Compute 95% confidence interval using t-distribution
confidence = 0.95
t_crit = ss.t.ppf((1 + confidence) / 2, df=n - 1)

margin_of_error = t_crit * standard_error
ci_lower = sample_mean - margin_of_error
ci_upper = sample_mean + margin_of_error

# Display results
print(f"Sample mean price: {sample_mean:,.2f} AED")
print(f"95% Confidence Interval: ({ci_lower:,.2f}, {ci_upper:,.2f}) AED")
print(f"Sample size (n): {n}")
print(f"Population size (N): 875,000")


Sample mean price: 2,028,387.53 AED
95% Confidence Interval: (1,990,298.08, 2,066,476.99) AED
Sample size (n): 3961
Population size (N): 875,000


## Show the Correlation Between Bedrooms and Bathrooms

This may require data preparation.

<b> State your answer, anlong with a statement of what it means in reality. (e.g. if I see correlation of .02 between 'shoe size' and 'IQ', I would say that the two appear to be unrelated, and knowing one tells us nothing about the other)</b>

#### Data Preparation

In [54]:
# Correlation Between Bedrooms and Bathrooms

# As the bedrooms and bathrooms were stored as object. So converting them to numeric and drop any missing values.
data_uae['bedrooms'] = pd.to_numeric(data_uae['bedrooms'], errors='coerce')
data_uae['bathrooms'] = pd.to_numeric(data_uae['bathrooms'], errors='coerce')

# Drop rows with missing bedrooms or bathrooms
corr_data = data_uae[['bedrooms', 'bathrooms']].dropna()


#### Compute Correlation

In [55]:
# Calculate Pearson correlation
correlation = corr_data['bedrooms'].corr(corr_data['bathrooms'])
print(f"Correlation between bedrooms and bathrooms: {correlation:.2f}")

Correlation between bedrooms and bathrooms: 0.89


#### Correlation Between Bedrooms and Bathrooms
The Pearson correlation between bedrooms and bathrooms is 0.89, indicating a strong positive relationship. This means that properties with more bedrooms almost always have more bathrooms. Knowing the number of bedrooms gives a very good idea of how many bathrooms a property is likely to have.

## Multiple Groups

Based on this data, do any of the groups (furnished, unfurnished, or partially furnished) appear to be different in terms of price per square foot? State your answer and show evidence to support your answer.

In [56]:
# Go

# I think first we need to create a new column for price per square foot
data_uae['price_per_sqft'] = data_uae['price'] / data_uae['sq_ft']

# Check unique values in furnishing column
data_uae['furnishing'].value_counts()

# Compute mean price per sqft for each furnishing group
group_summary = data_uae.groupby('furnishing')['price_per_sqft'].agg(['count', 'mean', 'std']).reset_index()
group_summary


furnishing  count         mean          std
0         NO   2804  1557.762158   797.790718
1     PARTLY    351  1430.408732   655.264142
2        YES    804  1922.973949  1006.617184

### Furnishing vs Price per Square Foot

Based on the data, there are noticeable differences in price per square foot between the furnishing groups. Fully furnished units have the highest average price per square foot at approximately 1,923 AED, followed by unfurnished units at around 1,558 AED, and partially furnished units at about 1,430 AED. This suggests that furnishing has a strong influence on price per square foot, with fully furnished properties generally commanding higher prices. The standard deviations indicate that there is some variability within each group, but the overall trend is clear the more complete furnishing is associated with higher price per square foot.

## Store

In this scenario, we are looking to puchase a small retail store, but don't want to do so unless we project to make a profit within the first year.

This store is run by a real nerd, so they gave us some statistical data for us to estimate (the exact values are in the code cell below as variables):
<ul>
<li> The number of customers per day is described by a Poisson distribution with a rate of <b><i>customers_per_day</i></b> customers per day.</li>
<li> The buy rate (the percentage of customers that make a purchase) is <b><i>buy_rate</i></b>.</li>
<li> The amount each customer spends is described by a normal distribution with a mean of <b><i>mean_purchase_per_buyer</i></b> and a standard deviation of <b><i>std_dev_purchase_per_buyer</i></b>.</li>
<li> Each day, we need to take in at least <b><i>minimum_daily_income</i></b> to cover costs - to make it a 'good day'.</li>
<li> In a given year (we open every day), we must have over <b><i>good_day_minimum_count</i></b> good days for the year to be a success.</li>
</ul>

To determine if the store is a good investment, we will simulate the daily sales over the course of a year and see how often we meet or exceed the <b><i>minimum_daily_income</i></b> threshold. <b> Demonstrate how comfortable you are with the statement "we express this store to be a success", and show why in your calculations.</b> This is not something with one specific answer, there are many details that may vary somewhat. If it is hard, I recommend starting with a very simple model - can you simulate one customer? Can we simulate one day of customers? Can we simulate one year of days? <b>Build a simple model and make it more accurate incrementally</b>, doing it all at once can be hard, doing it piece by piece is easier. 

<b>Overall, for grades, this will be roughly logarithmicly evaluated - if you get something ok, you'll get points. The better it is the more points, but as you go up, the curve flattens. The most value is in getting a simple, working version. </b>

<b>tl;dr:</b> We have a store with customers and purchases defined by these distributions. Simulate a year of operations and deterimine the number of dayas in that year the store made over the cutoff - if it did, that's a success, if not, it's a failure. 

In [58]:
customers_per_day = 200
buy_rate = 0.22
mean_purchase_per_buyer = 125
stddev_purchase_per_buyer = 30
minimum_daily_income = 4000
days_in_year = 365

good_day_minimum_rate = .10
good_day_minimum_count = days_in_year - (int(days_in_year * good_day_minimum_rate))

# Get simulating!

# Set seed for reproducibility
np.random.seed(42)

# Simulate number of customers per day (Poisson)
daily_customers = np.random.poisson(lam=customers_per_day, size=days_in_year)

# Simulate number of buyers per day (Binomial)
daily_buyers = np.random.binomial(n=daily_customers, p=buy_rate)

# Simulate daily income using vectorized normal draws
daily_income = np.array([
    np.random.normal(loc=mean_purchase_per_buyer, scale=stddev_purchase_per_buyer, size=buyers).sum()
    if buyers > 0 else 0
    for buyers in daily_buyers
])

good_days = np.sum(daily_income >= minimum_daily_income)

store_success = good_days >= good_day_minimum_count

# Results
print(f"Number of good days: {good_days} / {days_in_year}")
print(f"Required for success: {good_day_minimum_count} days")
print(f"Store is a success? {'Yes' if store_success else 'No'}")



Number of good days: 347 / 365
Required for success: 329 days
Store is a success? Yes


#### Store Simulation Analysis
After simulating a full year of store operations, the store had 347 good days where the daily income exceeded the minimum required of 4,000 AED. Since this is greater than the required threshold of 329 good days, we can conclude that the store is likely to be profitable within its first year. This demonstrates that, under the given assumptions about customer arrivals, purchase behavior, and daily spending, the store consistently meets or exceeds its minimum income target, supporting the statement that we expect this store to be a success.